### PR TITLE
Improve linspace decomposition and remove its lowering

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91625
* __->__ #91621

The code produced by the lowering and the decomposition is now the same
modulo a casting to `float32`. This casting is necessary as otherwise
the tests do not pass due to accuracy errors. We prefer accuracy over
speed here, given that this is an associative scan, and thus it's prone
to numerical errors.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by lezcano** [comment]
The test script
```python
import torch

def f():
    return torch.linspace(3, 10, steps=5, device="cuda")

compiled_model = torch.compile(f)
compiled_model()
```

generates in master
```python
    xnumel = 5
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = 1.75
    tmp1 = x0
    tmp2 = tmp0 * tmp1
    tmp3 = 3
    tmp4 = tmp2 + tmp3
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp4, xmask)
```

and in this PR
```python
    xnumel = 5
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.reshape(tl.arange(0, XBLOCK), [XBLOCK])
    xmask = xindex < xnumel
    x0 = xindex
    tmp0 = x0
    tmp1 = tmp0.to(tl.float32)
    tmp2 = 1.75
    tmp3 = tmp1 * tmp2
    tmp4 = 3
    tmp5 = tmp3 + tmp4
    tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp5, xmask)
```
The only difference is that the cast to float32 is explicit rather than implicit. This cast should probably be removed by a finer version of @eellison's strength reduction algorithm.

Note that the lowering had severe errors on low-precision dtypes. The current implementation fixes that.

@jancervenka have a look in case you want to have a go at removing some other lowerings while keeping the generated code the same.

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -4328,20 +4328,21 @@ def linspace(
...
-    double_dtype = torch.complex128 if utils.is_complex_dtype(dtype) else torch.float64
-    rg = _maybe_convert_to_dtype(rg, double_dtype)  # type: ignore[assignment]
-    cast = partial(torch.full, (1,), dtype=double_dtype, **factory_kwargs)
-    out = torch.lerp(cast(start), cast(end), rg)
+    computation_dtype, _ = utils.reduction_dtypes(
+        rg, REDUCTION_OUTPUT_TYPE_KIND.SAME, dtype_red
+    )
+
+    cast = partial(torch.full, (), dtype=computation_dtype, **factory_kwargs)
+    out = rg * cast((end - start) / (steps - 1)) + cast(start)
```
Comment (peterbell10) [inline torch/_refs/__init__.py:4345]: Should this not use the algorithm from ATen:

```
where(ind < halfway, start + step * rg, end - step * (steps - rg - 1))
```

This should get the end points of the range exactly.

Comment (lezcano) [reply torch/_refs/__init__.py:4345]: Sure thing. Let's kill two birds with one stone.

Comment (lezcano) [reply torch/_refs/__init__.py:4345]: https://github.com/pytorch/pytorch/pull/91625

Comment (ngimel) [reply torch/_refs/__init__.py:4345]: Why do we need to cast `end` and `start` to tensor? Inductor remedies this, but it still looks unnecessary?

Comment (lezcano) [reply torch/_refs/__init__.py:4345]: Sure, let me cast `rg` instead.

**Comment by lezcano** [comment]
Yeah, sorry for not updating the PR. Now it should be fine. Also, note that the next PR in the stack solves the issue with the ends of linspace.

**Comment by lezcano** [comment]
@ngimel now the two PRs are ready for review.

**Review by ngimel — APPROVED** [review]
